File: Easy/Print matrix in diagonal pattern/print-matrix-in-diagonal-pattern.cpp
//{ Driver Code Starts

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
/*You are required to complete this method */

class Solution{
  public:
    vector<int> matrixDiagonally(vector<vector<int>>&mat)
    {
        int n = mat.size();
        vector<int> adj[n+n-1];
        
        for(int i=0; i<n; i++) {
            for(int j=0; j<n; j++) {
                adj[i + j].push_back(mat[i][j]);
            }
        }
        
        vector<int> res;
        for(int i=0; i<n+n-1; i++) {
            if(i % 2 == 0)
                reverse(adj[i].begin(), adj[i].end());
            for(auto x: adj[i])
                res.push_back(x);
        }
        
        return res;
    }
};


//{ Driver Code Starts.
int main()
{
    int T;
    cin>>T;
  
    while(T--)
    {
        int n;
        int k=0;
        //cin>>k;
        cin>>n;
        vector<vector<int>>mat(n, vector<int>(n));
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                cin >> mat[i][j];
            }
        }
        Solution obj;
        vector<int>ans = obj.matrixDiagonally(mat);
        for(auto i: ans)cout << i << " ";
        cout << "\n";
        
       
    }
}
// } Driver Code Ends

File: Medium/Print matrix in diagonal pattern/print-matrix-in-diagonal-pattern.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            int[][] mat = new int[n][n];
            String[] S = br.readLine().trim().split(" ");
            int i = 0;
            int j = 0;
            for(int k = 0; k < S.length; k++){
                mat[i][j] = Integer.parseInt(S[k]);
                j++;
                if(j == n){
                    i++;
                    j = 0;
                }
            }
            Solution obj = new Solution();
            int[] ans = obj.matrixDiagonally(mat);
            for(int it = 0; it < ans.length; it++){
                System.out.print(ans[it] + " ");
            }
            System.out.println();
        }
    }
}

// } Driver Code Ends


class Solution {
    public int[] matrixDiagonally(int[][] arr) {
        
        int n = arr.length;
        int[] res = new int[n * n];
        int p = 0;
    
        int i = 0, j = 0;
    
        while (i < n && j < n) {
            // Traverse diagonally up
            while (i >= 0 && j < n) {
                res[p++] = arr[i--][j++];
            }
            
            // Adjust i and j for next iteration
            if (i < 0 && j < n) {
                i = 0;
            } else if (j == n) {
                i += 2;
                j--;
            }
            
            // Traverse diagonally down
            while (i < n && j >= 0) {
                res[p++] = arr[i++][j--];
            }
            
            // Adjust i and j for next iteration
            if (j < 0 && i < n) {
                j = 0;
            } else if (i == n) {
                j += 2;
                i--;
            }
        }
    
        return res;

    }
}